skytap/environment.go

```diff
@@ -272,10 +283,150 @@ func (s *EnvironmentsServiceClient) Delete(ctx context.Context, id string) error
 	if err != nil {
 		return err
 	}
-	_, err = s.client.do(ctx, req, nil)
+	_, err = s.client.do(ctx, req, nil, envRunStateNotBusy(id), nil)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func (payload *CreateEnvironmentRequest) compare(ctx context.Context, c *Client, v interface{}, state *environmentVMRunState) (string, bool) {
+	if envOriginal, ok := v.(*Environment); ok {
+		env, err := c.Environments.Get(ctx, *envOriginal.ID)
+		if err != nil {
+			return requestNotAsExpected, false
+		}
+		logEnvironmentStatus(env)
+		log.Printf("[DEBUG] SDK environment runstate after create (%s)\n", *env.Runstate)
+		if *env.Runstate != EnvironmentRunstateBusy {
+			return "", true
+		}
+		return "environment not ready", false
+	}
+	log.Printf("[ERROR] SDK environment comparison not possible on (%v)\n", v)
+	return requestNotAsExpected, false
+}
+
+func (payload *UpdateEnvironmentRequest) compare(ctx context.Context, c *Client, v interface{}, state *environmentVMRunState) (string, bool) {
+	if envOriginal, ok := v.(*Environment); ok {
+		env, err := c.Environments.Get(ctx, *envOriginal.ID)
+		if err != nil {
+			return requestNotAsExpected, false
+		}
+		logEnvironmentStatus(env)
+		actual := payload.buildUpdateRequestFromVM(env)
+		if payload.string() == actual.string() {
+			return "", true
+		}
+		return "environment not ready", false
+	}
+	log.Printf("[ERROR] SDK environment comparison not possible on (%v)\n", v)
+	return requestNotAsExpected, false
+}
+
+func (payload *UpdateEnvironmentRequest) buildUpdateRequestFromVM(env *Environment) UpdateEnvironmentRequest {
```

Comment:
why do we need all this if then, can not have nill in the Update Request? why the method name is `fromVM` when retrieves and environment?

Reply:
I think the best answer is because I am using a string to compare the objects.